src/platform/ESP32/NetworkCommissioningWiFiDriver.cpp

```diff
@@ -0,0 +1,327 @@
+        scan_config.ssid = WiFiSSIDStr;
+        err              = esp_wifi_scan_start(&scan_config, false);
+    }
+    else
+    {
+        err = esp_wifi_scan_start(NULL, false);
+    }
+    if (err != ESP_OK)
+    {
+        return chip::DeviceLayer::Internal::ESP32Utils::MapError(err);
+    }
+    return CHIP_NO_ERROR;
+}
+
+bool GetWiFiScanRespFromAPRecord(wifi_ap_record_t * ap_record, WiFiScanResponse & scanResponse)
+{
+    scanResponse.security = ap_record->authmode;
+    scanResponse.ssidLen  = strnlen((const char *) ap_record->ssid, DeviceLayer::Internal::kMaxWiFiSSIDLength);
+    memcpy((char *) scanResponse.ssid, ap_record->ssid, scanResponse.ssidLen);
+    memcpy(scanResponse.bssid, ap_record->bssid, 6);
+    scanResponse.channel  = ap_record->primary;
+    scanResponse.wiFiBand = DeviceLayer::NetworkCommissioning::WiFiBand::k2g4; // ESP32 platform only support 2.4G Hz WiFi Band
+    scanResponse.rssi     = ap_record->rssi;
+    return true;
+}
+
+void ESPWiFiDriver::OnScanWiFiNetworkDone()
+{
+    uint16_t ap_number;
+    esp_wifi_scan_get_ap_num(&ap_number);
+    if (!ap_number)
+    {
+        ChipLogProgress(DeviceLayer, "No AP found");
+        if (mpScanCallback != nullptr)
+        {
+            mpScanCallback->OnFinished(Status::kSuccess, CharSpan(), nullptr);
+            mpScanCallback = nullptr;
+        }
+        return;
+    }
+    wifi_ap_record_t * ap_list_buffer;
+    ap_list_buffer = (wifi_ap_record_t *) malloc(ap_number * sizeof(wifi_ap_record_t));
+    if (ap_list_buffer == NULL)
+    {
+        ChipLogError(DeviceLayer, "can't malloc memory for ap_list_buffer");
+        if (mpScanCallback)
+        {
+            mpScanCallback->OnFinished(Status::kUnknownError, CharSpan(), nullptr);
+            mpScanCallback = nullptr;
+        }
+        return;
+    }
+    if (esp_wifi_scan_get_ap_records(&ap_number, (wifi_ap_record_t *) ap_list_buffer) == ESP_OK)
+    {
+        DeviceLayer::SystemLayer().ScheduleLambda([ap_number, ap_list_buffer]() {
+            ESPScanResponseIterator iter(ap_number, const_cast<wifi_ap_record_t *>(ap_list_buffer));
+            if (GetInstance().mpScanCallback)
+            {
+                GetInstance().mpScanCallback->OnFinished(Status::kSuccess, CharSpan(), &iter);
+                GetInstance().mpScanCallback = nullptr;
+                free(ap_list_buffer);
+            }
+        });
+    }
+    else
+    {
+        ChipLogError(DeviceLayer, "can't get ap_records ");
+        if (mpScanCallback)
+        {
+            mpScanCallback->OnFinished(Status::kUnknownError, CharSpan(), nullptr);
+            mpScanCallback = nullptr;
+        }
+    }
+}
+
+void ESPWiFiDriver::ScanNetworks(ByteSpan ssid, WiFiDriver::ScanCallback * callback)
+{
+    CHIP_ERROR err = StartScanWiFiNetworks(ssid);
+    mpScanCallback = callback;
+    if (err != CHIP_NO_ERROR)
+    {
+        mpScanCallback = nullptr;
+        callback->OnFinished(Status::kUnknownError, CharSpan(), nullptr);
```

Comment:
this method seems to assume callback is never null. Maybe add a check at the top?

Reply:
Add a check at the top